cluster_library.c

```diff
@@ -3105,22 +3157,31 @@ zend_string *cluster_hash_seeds(zend_string **seeds, uint32_t count) {
 }
 
 PHP_REDIS_API redisCachedCluster *cluster_cache_load(zend_string *hash) {
+    redisCachedCluster *cc;
     zend_resource *le;
 
     /* Look for cached slot information */
     le = zend_hash_find_ptr(&EG(persistent_list), hash);
+    if (le == NULL)
+        return NULL;
 
-    if (le != NULL) {
-        /* Sanity check on our list type */
-        if (le->type == le_cluster_slot_cache) {
-            /* Success, return the cached entry */
-            return le->ptr;
-        }
+    if (le->type != le_cluster_slot_cache) {
         php_error_docref(0, E_WARNING, "Invalid slot cache resource");
+        return NULL;
     }
 
-    /* Not found */
-    return NULL;
+    cc = le->ptr;
+    /* Short circuit if it should be expired */
+    if (cc->expiry != 0 && cc->expiry <= time(NULL))
+        return NULL;
```

Comment:
I guess we need to remove element from `EG(persistent_list)` in this case and when generation mismatch

Reply:
I think we get that for free with our dtor callback.

I can check though.

Reply:
You're right. I thought that the destruction would happen automatically when the cache is replaced like other zend hash tables do.